lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+        }
+        r.Set(v);
+        setXformOpOrder();
+    }
+    else {
+        r = getOp(NdxRotate);
+        cvt = getCvtRotXYZToAttrFn(r.GetOpName());
+    }
+
+    return std::make_shared<UsdRotateOpUndoableCmd>(
+        v, path(), r, cvt, UsdTimeCode::Default());
+}
+
+Ufe::ScaleUndoableCommand::Ptr UsdTransform3dMayaXformStack::scaleCmd(double x, double y, double z)
+{
+    // If there is no scale transform op yet, create it.
+    UsdGeomXformOp s;
+    GfVec3f v(x, y, z);
+    if (!hasOp(NdxScale)) {
+        InTransform3dChange guard(path());
+        s = _xformable.AddScaleOp();
+        if (!TF_VERIFY(s)) {
+            return nullptr;
+        }
+        s.Set(v);
+        setXformOpOrder();
+    }
+    else {
+        s = getOp(NdxScale);
+    }
+
+    return std::make_shared<UsdVecOpUndoableCmd<GfVec3f>>(
+        v, path(), s, UsdTimeCode::Default());
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::rotatePivotCmd(double x, double y, double z)
+{
+    return pivotCmd(UsdMayaXformStackTokens->rotatePivot, x, y, z);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::rotatePivot() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, UsdMayaXformStackTokens->rotatePivot));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::scalePivotCmd(double x, double y, double z)
+{
+    return pivotCmd(UsdMayaXformStackTokens->scalePivot, x, y, z);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::scalePivot() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, UsdMayaXformStackTokens->scalePivot));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::translateRotatePivotCmd(
+    double x, double y, double z)
+{
+    auto opSuffix = UsdMayaXformStackTokens->rotatePivotTranslate;
+    auto attrName = UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, opSuffix);
+    return setVector3dCmd(GfVec3f(x, y, z), attrName, opSuffix);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::rotatePivotTranslation() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate,
+        UsdMayaXformStackTokens->rotatePivotTranslate));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::translateScalePivotCmd(
+    double x, double y, double z)
+{
+    auto opSuffix = UsdMayaXformStackTokens->scalePivotTranslate;
+    auto attrName = UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, opSuffix);
+    return setVector3dCmd(GfVec3f(x, y, z), attrName, opSuffix);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::scalePivotTranslation() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate,
+        UsdMayaXformStackTokens->scalePivotTranslate));
+}
+
+template<class V>
+Ufe::Vector3d UsdTransform3dMayaXformStack::getVector3d(
```

Comment:
Generic accessor used for translation, rotate pivot, scale pivot, rotate pivot compensation, and scale pivot compensation.